### PR TITLE
Sync Upstream

### LINKED ISSUE #79661: thread 'rustc' panicked at 'called `Option::unwrap()` on a `None` value', compiler/rustc_middle/src/ty/query/mod.rs:235:5
### Code

This happened while working on cg_clif.

### Meta

`rustc --version --verbose`:
```
rustc 1.50.0-nightly (f4db9ffb2 2020-12-02)
binary: rustc
commit-hash: f4db9ffb22dfcb702dbdb2e0607cb91791866b57
commit-date: 2020-12-02
host: x86_64-unknown-linux-gnu
release: 1.50.0-nightly
```

### Error output

```
   Compiling rustc_codegen_cranelift v0.1.0 (/home/bjorn/Documenten/cg_clif3)
thread 'rustc' panicked at 'called `Option::unwrap()` on a `None` value', compiler/rustc_middle/src/ty/query/mod.rs:235:5
stack backtrace:
   0: rust_begin_unwind
             at /rustc/f4db9ffb22dfcb702dbdb2e0607cb91791866b57/library/std/src/panicking.rs:493:5
   1: core::panicking::panic_fmt
             at /rustc/f4db9ffb22dfcb702dbdb2e0607cb91791866b57/library/core/src/panicking.rs:92:14
   2: core::panicking::panic
             at /rustc/f4db9ffb22dfcb702dbdb2e0607cb91791866b57/library/core/src/panicking.rs:50:5
   3: rustc_middle::ty::query::try_load_from_on_disk_cache
   4: rustc_query_system::dep_graph::graph::DepGraph<K>::exec_cache_promotions
   5: rustc_middle::dep_graph::<impl rustc_query_system::dep_graph::DepKind for rustc_middle::dep_graph::dep_node::DepKind>::with_deps
   6: rustc_incremental::persist::save::save_in
   7: rustc_data_structures::sync::join
   8: rustc_middle::dep_graph::<impl rustc_query_system::dep_graph::DepKind for rustc_middle::dep_graph::dep_node::DepKind>::with_deps
   9: rustc_incremental::persist::save::save_dep_graph
  10: rustc_codegen_ssa::base::finalize_tcx
  11: rustc_codegen_ssa::base::codegen_crate
  12: <rustc_codegen_llvm::LlvmCodegenBackend as rustc_codegen_ssa::traits::backend::CodegenBackend>::codegen_crate
  13: rustc_interface::queries::Queries::ongoing_codegen
  14: rustc_interface::queries::<impl rustc_interface::interface::Compiler>::enter
  15: rustc_span::with_source_map
  16: scoped_tls::ScopedKey<T>::set
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.50.0-nightly (f4db9ffb2 2020-12-02) running on x86_64-unknown-linux-gnu

note: compiler flags: -Z run_dsymutil=no -C opt-level=3 -C embed-bitcode=no -C incremental -C link-arg=-Wl,-rpath=$ORIGIN/../lib --crate-type dylib

note: some of the compiler flags provided by cargo are hidden

query stack during panic:
end of query stack
thread '<unnamed>' panicked at 'assertion failed: `(left == right)`
  left: `LLVMing`,
 right: `Codegenning`', /rustc/f4db9ffb22dfcb702dbdb2e0607cb91791866b57/compiler/rustc_codegen_ssa/src/back/write.rs:1425:21
stack backtrace:
   0: rust_begin_unwind
             at /rustc/f4db9ffb22dfcb702dbdb2e0607cb91791866b57/library/std/src/panicking.rs:493:5
   1: core::panicking::panic_fmt
             at /rustc/f4db9ffb22dfcb702dbdb2e0607cb91791866b57/library/core/src/panicking.rs:92:14
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.50.0-nightly (f4db9ffb2 2020-12-02) running on x86_64-unknown-linux-gnu

note: compiler flags: -Z run_dsymutil=no -C opt-level=3 -C embed-bitcode=no -C incremental -C link-arg=-Wl,-rpath=$ORIGIN/../lib --crate-type dylib

note: some of the compiler flags provided by cargo are hidden

query stack during panic:
end of query stack
error: could not compile `rustc_codegen_cranelift`

To learn more, run the command again with --verbose.
```

### PR DESCRIPTION
**Pull request by vext01**
Fixes [the incremental compilation issue](https://github.com/rust-lang/rust/issues/79661).

### CODE REVIEW

**Comment by ltratt** [comment]
This is a sync I've been looking forward to!

bors r+

**Comment by ltratt** [comment]
Those failures look a bit scary to me, so I'll leave this one to you to look at (I was hoping for the tmp dir problem!).

**Comment by bjorn3** [comment]
https://github.com/rust-lang/rust/blob/80cc2ecf1070728a7d006026e0d532d9fc48621c/library/std/src/sys/unix/kernel_copy/tests.rs#L78

https://github.com/rust-lang/rust/blob/80cc2ecf1070728a7d006026e0d532d9fc48621c/library/std/src/sys/unix/kernel_copy/tests.rs#L107

https://github.com/rust-lang/rust/blob/80cc2ecf1070728a7d006026e0d532d9fc48621c/library/std/src/sys/unix/kernel_copy/tests.rs#L138

Can you take a look at `/tmp` to see if there are any files used by these tests with the wrong permissions? These tests don't seem to cleanup afterwards.

Edit: "I was hoping for the tmp dir problem!" yeah, that was also my guess :)

**Comment by ltratt** [comment]
@bjorn3 Oof, yes, I can see those dirs. I'm ... surprised ... that such tests keep getting merged in! I've deleted them so let's see what happens. Thanks for the pointer :)

bors r+

**Comment by vext01** [comment]
It's this again: https://github.com/rust-lang/rust/pull/79375

@bjorn3 is there any way we can accelerate merging that? We keep hitting this.

We could cherry pick my fixes, but I'd rather it were fixed upstream.

**Comment by ltratt** [comment]
I'm an idiot -- I was looking for something very different in the output, but clearly I wasn't thinking properly!

**Comment by ltratt** [comment]
Hmm, let me try again.

bors r+

**Comment by ltratt** [comment]
I'm going to assume that we're running the tests twice, or similar, and that the directories are being recreated *during* the run. @vext01 what do you want to do?

**Comment by vext01** [comment]
> @vext01 what do you want to do?

Bjorn has kindly started the merge for the fix.

In the meantime, we ought to be able to get this through, it's just racy.

I've killed the directories in /tmp. Please try again.
